File: packages/parser/src/diagnostics.ts
export enum DiagnosticCode {
  ParseError = 'ttr/parse-error',
  ParseRecoveryInfo = 'ttr/parse-recovery-info',
  UnknownProperty = 'ttr/unknown-property',
  UnresolvedReference = 'ttr/unresolved-reference',
  DuplicateDefinition = 'ttr/duplicate-definition',
  RequiredPropertyMissing = 'ttr/required-property-missing',
  InvalidType = 'ttr/invalid-type',
  EntityAttributeNotFound = 'ttr/entity-attribute-not-found',
  PrimaryKeyColumnNotFound = 'ttr/primary-key-column-not-found',
  WrongFileKind = 'ttr/wrong-file-kind',
  UnimportedReference = 'ttr/unimported-reference',
  UnusedImport = 'ttr/unused-import',
  WildcardWithNoMatches = 'ttr/wildcard-with-no-matches',
  DuplicateImport = 'ttr/duplicate-import',
  CircularPackageDependency = 'ttr/circular-package-dependency',
  PackageDeclarationMismatch = 'ttr/package-declaration-mismatch',
  PackagePrefixDivergence = 'ttr/package-prefix-divergence',
  MissingPackageDeclaration = 'ttr/missing-package-declaration',
  // Packages & Domains PD3 — `.ttrd` domain resolution.
  DomainMemberNotFound = 'ttr/domain-member-not-found',
  DomainEmpty = 'ttr/domain-empty',
  DuplicateDomain = 'ttr/duplicate-domain',
  DomainRedundantMember = 'ttr/domain-redundant-member',
  AmbiguousReference = 'ttr/ambiguous-reference',
  GraphObjectNotFound = 'ttr/graph-object-not-found',
  GraphLayoutStaleNode = 'ttr/graph-layout-stale-node',
  GraphObjectsEmpty = 'ttr/graph-objects-empty',
  GraphNameMismatch = 'ttr/graph-name-mismatch',
  FileOrdering = 'ttr/file-ordering',
  FuzzyWithoutSearchable = 'ttr/fuzzy-without-searchable',
  DuplicateSearchProperty = 'ttr/duplicate-search-property',
  DuplicateBinding = 'ttr/duplicate-binding',
  // embedded-sql (DESIGN §5/§6): tagged-block tag resolution.
  UnknownLanguageTag = 'ttr/unknown-language-tag',
  LanguageTagMismatch = 'ttr/language-tag-mismatch',
  DeprecatedLanguageProperty = 'ttr/deprecated-language-property',
  // embedded-sql (DESIGN §12.8 / contracts §5.1): SQL reference resolution
  // against the TTR `db` symbol table. Best-effort (warnings); see §3.4/§3.5.
  SqlUnknownTable = 'sql-unknown-table',
  SqlUnknownColumn = 'sql-unknown-column',
  SqlAmbiguousColumn = 'sql-ambiguous-column',
  SqlColumnNotOnAlias = 'sql-column-not-on-alias',
  SqlUndeclaredParam = 'sql-undeclared-param',
  SqlUnusedParam = 'sql-unused-param',
}

export enum DiagnosticSeverity {
  Error = 'error',
  Warning = 'warning',
  Information = 'information',
  Hint = 'hint',
}

export interface DiagnosticEntry {
  code: DiagnosticCode;
  message: string;
  severity: DiagnosticSeverity;
}

File: packages/python/ttr-parser/src/ttr_parser/diagnostics.py
"""Diagnostic codes and severity shared across the parser and semantics layers.

The `DiagnosticCode` enum's `.value` strings are part of the public contract —
they match the Kotlin `DiagnosticCode.id` (and the TS `DiagnosticCode.*` enum
literal strings) byte-for-byte so the conformance diagnostic dump (§5.1) is
interchangeable across the three runtimes.

The parser emits the parse-level codes (`PARSE_ERROR`, `PARSE_RECOVERY_INFO`).
The semantics layer (P4) emits the resolution/validation codes. Both layers
import from this single canonical enum.
"""

from __future__ import annotations

from enum import Enum


class DiagnosticCode(Enum):
    PARSE_ERROR = "ttr/parse-error"
    PARSE_RECOVERY_INFO = "ttr/parse-recovery-info"
    UNKNOWN_PROPERTY = "ttr/unknown-property"
    UNRESOLVED_REFERENCE = "ttr/unresolved-reference"
    DUPLICATE_DEFINITION = "ttr/duplicate-definition"
    REQUIRED_PROPERTY_MISSING = "ttr/required-property-missing"
    INVALID_TYPE = "ttr/invalid-type"
    ENTITY_ATTRIBUTE_NOT_FOUND = "ttr/entity-attribute-not-found"
    PRIMARY_KEY_COLUMN_NOT_FOUND = "ttr/primary-key-column-not-found"
    WRONG_FILE_KIND = "ttr/wrong-file-kind"
    UNIMPORTED_REFERENCE = "ttr/unimported-reference"
    UNUSED_IMPORT = "ttr/unused-import"
    WILDCARD_WITH_NO_MATCHES = "ttr/wildcard-with-no-matches"
    DUPLICATE_IMPORT = "ttr/duplicate-import"
    CIRCULAR_PACKAGE_DEPENDENCY = "ttr/circular-package-dependency"
    PACKAGE_DECLARATION_MISMATCH = "ttr/package-declaration-mismatch"
    MISSING_PACKAGE_DECLARATION = "ttr/missing-package-declaration"
    AMBIGUOUS_REFERENCE = "ttr/ambiguous-reference"
    GRAPH_OBJECT_NOT_FOUND = "ttr/graph-object-not-found"
    GRAPH_LAYOUT_STALE_NODE = "ttr/graph-layout-stale-node"
    GRAPH_OBJECTS_EMPTY = "ttr/graph-objects-empty"
    GRAPH_NAME_MISMATCH = "ttr/graph-name-mismatch"
    FILE_ORDERING = "ttr/file-ordering"
    FUZZY_WITHOUT_SEARCHABLE = "ttr/fuzzy-without-searchable"
    DUPLICATE_SEARCH_PROPERTY = "ttr/duplicate-search-property"
    DUPLICATE_BINDING = "ttr/duplicate-binding"
    UNKNOWN_LANGUAGE_TAG = "ttr/unknown-language-tag"
    LANGUAGE_TAG_MISMATCH = "ttr/language-tag-mismatch"
    DEPRECATED_LANGUAGE_PROPERTY = "ttr/deprecated-language-property"

    def __str__(self) -> str:
        return self.value


class DiagnosticSeverity(Enum):
    ERROR = "Error"
    WARNING = "Warning"
    INFORMATION = "Information"
    HINT = "Hint"

    def __str__(self) -> str:
        return self.value